### PR TITLE
HHH-13169 - use exact table name for multitable update queries instea…

### PR DESCRIPTION
**Pull request by Mobe91**
…d of table alias

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElementType.java [hunk 1]
```diff
@@ -456,11 +456,15 @@ public Type getPropertyType(String propertyName, String propertyPath) {
...
+			// for update queries, we must only use the alias in the top level where clause
 			// B) otherwise, we need to use the persister's
 			// table name as the column qualification
 			// 2) otherwise (not correlated), use the given alias
 			if ( isCorrelation() ) {
-				if ( isMultiTable() || isInsertQuery() ) {
+				if (isMultiTable() && (!isUpdateQuery() || fromElement.getWalker().getCurrentTopLevelClauseType() == HqlSqlTokenTypes.WHERE)) {
+					return propertyMapping.toColumns(tableAlias, path);
+				}
+				else if (isInsertQuery()) {
```
Comment (JWGmeligMeyling) [inline hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElementType.java]: Shouldn't this still check for `isMultiTable()`?

Comment (Mobe91) [reply hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElementType.java]: You mean leave it like `else if ( isMultiTable() || isInsertQuery() )`? Then you would get the same erroneous behavior as before.

File: hibernate-core/src/test/java/org/hibernate/test/hql/UpdateJoinedSubclassCorrelationTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import javax.persistence.*;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+
+import java.util.List;
+
+import static javax.persistence.InheritanceType.JOINED;
+import static org.hibernate.testing.transaction.TransactionUtil.doInJPA;
+
+@TestForIssue(jiraKey = "")
```
Comment (gsmet) [inline hibernate-core/src/test/java/org/hibernate/test/hql/UpdateJoinedSubclassCorrelationTest.java]: Please add the JIRA issue here.

File: hibernate-core/src/test/java/org/hibernate/test/hql/UpdateJoinedSubclassCorrelationTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		} );
+
+		doInJPA( this::entityManagerFactory, entityManager -> {
+			// DO NOT CHANGE this query: it used to trigger a very specific bug caused
+			// by the root table alias being added to the generated subquery instead of the table name
+			String d = "update SubMaster m set name = (select 'test' from Detail d where d.master = m)";
+			Query del = entityManager.createQuery( d );
+			del.executeUpdate();
+
+			// so check for exactly one master after deletion
```
Comment (gsmet) [inline hibernate-core/src/test/java/org/hibernate/test/hql/UpdateJoinedSubclassCorrelationTest.java]: This comment does not seem related to what the test does.

File: hibernate-core/src/test/java/org/hibernate/test/hql/UpdateJoinedSubclassCorrelationTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			Detail d11 = new Detail( 10, m1 );
+			entityManager.persist( d11 );
+		} );
+
+		doInJPA( this::entityManagerFactory, entityManager -> {
+			// DO NOT CHANGE this query: it used to trigger a very specific bug caused
+			// by the root table alias being added to the generated subquery instead of the table name
+			String d = "update SubMaster m set name = (select 'test' from Detail d where d.master = m)";
+			Query del = entityManager.createQuery( d );
+			del.executeUpdate();
```
Comment (gsmet) [inline hibernate-core/src/test/java/org/hibernate/test/hql/UpdateJoinedSubclassCorrelationTest.java]: Looks like you copied a test doing deletes `d`, `del`. Maybe you could change that?

**Comment by gbadner** [comment]
The fix looks OK to me.

I tried some related queries queries, and I think there may still be a problem. If so, should they be dealt with in the same jira issue, or with a new one?

The following query is valid and generates the proper SQL:

`update SubMaster m set name = (select 'test' from Detail d where d.master.id = m.id)`

I am not 100% sure if the following query is a valid according to JPA 2.2:: 

`update SubMaster set name = (select 'test' from Detail d where d.master = SubMaster)`

but the generated SQL does not look correct:

`update SubMaster set name=(select 'test' from Detail updatejoin1_ where updatejoin1_.master_id=1) where ( id ) IN ( select id from HT_SubMaster)`

Specifically, `updatejoin1_.master_id=1` doesn't look correct. The test passes because the `SubMaster` ID is 1. If I change the ID to 2, it fails.

I think the following query is valid, but not 100% sure:

`update SubMaster set name = (select 'test' from Detail d where d.master.id = SubMaster.id)`

which fails:

 Invalid path: '1.id'
	at org.hibernate.hql.internal.ast.util.LiteralProcessor.lookupConstant(LiteralProcessor.java:111)
	at org.hibernate.hql.internal.ast.tree.DotNode.resolve(DotNode.java:214)
	at org.hibernate.hql.internal.ast.HqlSqlWalker.resolve(HqlSqlWalker.java:1045)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.expr(HqlSqlBaseWalker.java:1290)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.exprOrSubquery(HqlSqlBaseWalker.java:4706)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.comparisonExpr(HqlSqlBaseWalker.java:4177)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.logicalExpr(HqlSqlBaseWalker.java:2138)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.whereClause(HqlSqlBaseWalker.java:815)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.query(HqlSqlBaseWalker.java:609)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.newValue(HqlSqlBaseWalker.java:1243)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.assignment(HqlSqlBaseWalker.java:1056)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.setClause(HqlSqlBaseWalker.java:765)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.updateStatement(HqlSqlBaseWalker.java:381)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.statement(HqlSqlBaseWalker.java:269)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.analyze(QueryTranslatorImpl.java:272)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.doCompile(QueryTranslatorImpl.java:192)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.compile(QueryTranslatorImpl.java:144)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:118)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:78)
	at org.hibernate.engine.query.spi.QueryPlanCache.getHQLQueryPlan(QueryPlanCache.java:158)
	at org.hibernate.internal.AbstractSharedSessionContract.getQueryPlan(AbstractSharedSessionContract.java:595)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:704)
	at org.hibernate.internal.AbstractSessionImpl.createQuery(AbstractSessionImpl.java:23)
	at org.hibernate.test.hql.UpdateJoinedSubclassCorrelationTest.lambda$testJoinedSubclassUpdateWithCorrelation$1(UpdateJoinedSubclassCorrelationTest.java:39)
	at org.hibernate.testing.transaction.TransactionUtil.doInJPA(TransactionUtil.java:210)
	at org.hibernate.testing.transaction.TransactionUtil.doInJPA(TransactionUtil.java:251)
	at org.hibernate.test.hql.UpdateJoinedSubclassCorrelationTest.testJoinedSubclassUpdateWithCorrelation(UpdateJoinedSubclassCorrelationTest.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.hibernate.testing.junit4.ExtendedFrameworkMethod.invokeExplosively(ExtendedFrameworkMethod.java:45)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
19:33:09,655 ERROR ErrorTracker:62 -  right-hand operand of a binary operator was null
19:33:09,655 ERROR ErrorTracker:57 -  right-hand operand of a binary operator was null
 right-hand operand of a binary operator was null
	at org.hibernate.hql.internal.ast.tree.BinaryLogicOperatorNode.initialize(BinaryLogicOperatorNode.java:45)
	at org.hibernate.hql.internal.ast.HqlSqlWalker.prepareLogicOperator(HqlSqlWalker.java:1451)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.comparisonExpr(HqlSqlBaseWalker.java:4648)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.logicalExpr(HqlSqlBaseWalker.java:2138)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.whereClause(HqlSqlBaseWalker.java:815)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.query(HqlSqlBaseWalker.java:609)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.newValue(HqlSqlBaseWalker.java:1243)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.assignment(HqlSqlBaseWalker.java:1056)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.setClause(HqlSqlBaseWalker.java:765)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.updateStatement(HqlSqlBaseWalker.java:381)
	at org.hibernate.hql.internal.antlr.HqlSqlBaseWalker.statement(HqlSqlBaseWalker.java:269)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.analyze(QueryTranslatorImpl.java:272)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.doCompile(QueryTranslatorImpl.java:192)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.compile(QueryTranslatorImpl.java:144)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:118)
	at org.hibernate.engine.query.spi.HQLQueryPlan.<init>(HQLQueryPlan.java:78)
	at org.hibernate.engine.query.spi.QueryPlanCache.getHQLQueryPlan(QueryPlanCache.java:158)
	at org.hibernate.internal.AbstractSharedSessionContract.getQueryPlan(AbstractSharedSessionContract.java:595)
	at org.hibernate.internal.AbstractSharedSessionContract.createQuery(AbstractSharedSessionContract.java:704)
	at org.hibernate.internal.AbstractSessionImpl.createQuery(AbstractSessionImpl.java:23)
	at org.hibernate.test.hql.UpdateJoinedSubclassCorrelationTest.lambda$testJoinedSubclassUpdateWithCorrelation$1(UpdateJoinedSubclassCorrelationTest.java:39)
	at org.hibernate.testing.transaction.TransactionUtil.doInJPA(TransactionUtil.java:210)
	at org.hibernate.testing.transaction.TransactionUtil.doInJPA(TransactionUtil.java:251)
	at org.hibernate.test.hql.UpdateJoinedSubclassCorrelationTest.testJoinedSubclassUpdateWithCorrelation(UpdateJoinedSubclassCorrelationTest.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.hibernate.testing.junit4.ExtendedFrameworkMethod.invokeExplosively(ExtendedFrameworkMethod.java:45)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)

**Comment by gsmet** [comment]
@gbadner I would vote for merging this one and creating another JIRA. It looks unrelated (and possibly a rabbit hole).

**Comment by beikov** [comment]
AFAIK the erroneous cases are invalid JPQL, so IMO it should generate a parser error. Using the entity name `SubMaster` should resolve to a class/type literal which is not comparable to anything but a type expression.

+1 for getting this in @gsmet

**Comment by gbadner** [comment]
@beikov, can you quote from the spec about `SubMaster` resolving to a class/type literal? I'm trying to understand...

It's fine with me to merge this.

**Comment by gsmet** [comment]
OK, perfect, I squashed the 2 commits and merged.

Thanks everyone.
